os/fs/smartfs/smartfs_utils.c

```diff
@@ -1040,259 +1101,284 @@ int smartfs_finddirentry(struct smartfs_mountpt_s *fs, struct smartfs_entry_s *d
 #endif
-#endif							/* CONFIG_SMARTFS_ERASEDSTATE == 0xFF */
-
-	if (sectorno == 0xFFFF) {
-		/* Allocate a new sector for the file / dir */
-
-		ret = FS_IOCTL(fs, BIOC_ALLOCSECT, 0xFFFF);
-		if (ret < 0) {
-			goto errout;
+			RESET_ENTRY_FLAGS(&new_entry);
 		}
+		SET_ENTRY_FLAGS(new_entry.flags, mode, type);
+	}
 
-		nextsector = (uint16_t)ret;
-
-		/* Set the newly allocated sector's type (file or dir) */
-
-#ifdef CONFIG_SMARTFS_USE_SECTOR_BUFFER
-		if (sf) {
-			/* Using sector buffer and we have an open file context.  Just update
-			 * the sector buffer in the open file context.
-			 */
-
-			memset(sf->buffer, CONFIG_SMARTFS_ERASEDSTATE, fs->fs_llformat.availbytes);
-			chainheader = (struct smartfs_chain_header_s *)sf->buffer;
-			chainheader->type = SMARTFS_SECTOR_TYPE_FILE;
-			sf->bflags = SMARTFS_BFLAG_DIRTY | SMARTFS_BFLAG_NEWALLOC;
-		} else
-#endif
-		{
-			if ((type & SMARTFS_DIRENT_TYPE) == SMARTFS_DIRENT_TYPE_DIR) {
-				chainheader->type = SMARTFS_SECTOR_TYPE_DIR;
-			} else {
-				chainheader->type = SMARTFS_SECTOR_TYPE_FILE;
-			}
-
-			readwrite.count = 1;
-			readwrite.offset = offsetof(struct smartfs_chain_header_s, type);
-			readwrite.buffer = (uint8_t *)&chainheader->type;
-			readwrite.logsector = nextsector;
-			ret = FS_IOCTL(fs, BIOC_WRITESECT, (unsigned long)&readwrite);
-			if (ret < 0) {
-				fdbg("Error %d setting new sector type for sector %d\n", ret, nextsector);
-				goto errout;
-			}
+	if (newchain) {
+		/* We cannot read the new sector into fs->fs_rwbuffer as it is totally empty.
+		   MTD header will be written when MTD_WRITE is called for the sector 1st time.
+		   Chainheader is written along with the 1st entry.
+		   So currently, smart_readsector will fail to invalidate CRC on empty sector
+		 */
+		tmp_buf = (char *)kmm_malloc(entrysize + sizeof(struct smartfs_entry_header_s));
+		if (tmp_buf == NULL) {
+			fdbg("Unable to allocate memory\n");
+			return -ENOMEM;
 		}
+		memset(tmp_buf, CONFIG_SMARTFS_ERASEDSTATE, entrysize + sizeof(struct smartfs_entry_header_s));
+		entry = (struct smartfs_entry_header_s *)&tmp_buf[sizeof(struct smartfs_chain_header_s)];
+		chainheader = (struct smartfs_chain_header_s *)&tmp_buf[0];
+		chainheader->type = SMARTFS_SECTOR_TYPE_DIR;
 	} else {
-		/* Use the provided sector number */
-
-		nextsector = sectorno;
+		smartfs_setbuffer(&readwrite, psector, 0, fs->fs_llformat.availbytes, (uint8_t *)fs->fs_rwbuffer);
+		ret = FS_IOCTL(fs, BIOC_READSECT, (unsigned long)&readwrite);
+		if (ret < 0) {
+			fdbg("Unable to read sector %d for writing the new entry\n", psector);
+			return ret;
+		}
+		entry = (struct smartfs_entry_header_s *)&fs->fs_rwbuffer[offset];
 	}
 
-	/* Create the directory entry to be written in the parent's sector */
+	/* Populate the directory entry to be written in the parent's sector */
 
 #ifdef CONFIG_SMARTFS_ALIGNED_ACCESS
-	smartfs_wrle16(&entry->firstsector, nextsector);
+	smartfs_wrle16(&entry->firstsector, new_entry.firstsector);
```

Comment:
it's node duplicated with line 1381~1401?

Reply:
Yes `direntry` object is also populated with same attributes for returning the entry that is written.

Reply:
ah ok I confused for a while